Define devfile versioning and release process (#278)
devfile/api: docs/proposals/versioning-and-release.md

# Devfile Versioning and Release Process
This design document outlines the proposed versioning and release process for the Devfile spec. 

> This document summarizes parts from he Devfile API technical meeting slides presented by @davidfestal back in October, and I recommend having a read through of it: https://docs.google.com/presentation/d/1ohM1HzPB59a3ajvB7rVWJkKONLBAuT0mb5P20_A6vLs/edit#slide=id.g8fc722bef9_1_8

Author: Ah, must've just been shared with us (Devfile Services). I think @davidfestal can grant access?

> This document summarizes parts from he Devfile API technical meeting slides presented by @davidfestal back in October, and I recommend having a read through of it: https://docs.google.com/presentation/d/1ohM1HzPB59a3ajvB7rVWJkKONLBAuT0mb5P20_A6vLs/edit#slide=id.g8fc722bef9_1_8

Reviewer: typo: "he Devfile API..." should be "the Devfile API..."

Author: Updated. My `t` is broken on my laptop so I have to hit extra hard for characters to show up 😅

> This document summarizes parts from he Devfile API technical meeting slides presented by @davidfestal back in October, and I recommend having a read through of it: https://docs.google.com/presentation/d/1ohM1HzPB59a3ajvB7rVWJkKONLBAuT0mb5P20_A6vLs/edit#slide=id.g8fc722bef9_1_8

Reviewer 2: @amisevsk You should have access now.

## Versioning

The following sections outline how we version the Devfile Kubernetes API, as well as the JSON schema generated from the API.

### Kubernetes API
The Devfile Kubernetes API (defined in https://github.com/devfile/api/) is the single source of truth for the Devfile spec. It’s defined under `pkg/apis/` and contains the Go structs that are used by the devworkspace operator and devfile library. It's also where the Devfile JSON schema is generated from. 

**Versioning Scheme**: [Kubernetes](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning) (e.g. v1alpha1, v1beta2, v1, v2, etc)

> **Update Process**: 

Reviewer: Consider changing the order of "Update Process" and "When to Update". I feel more natural to read in that order.

Author: Sounds good to me, updated 👍 .

   1) Add a new folder in the source repository (`pkg/apis/workspaces` in `devfile/api`)

>    2) Add a schema and version in the CRD manifests

Reviewer: Should we provide the path to it?

Author: Sounds reasonable to me, updated.

   3) Generate the JSON schema from the API. New JSON schema will be located under `schemas/latest`.
   4) Update the devworkspace operator and devfile library to consume the Go structs in the new API version, as needed.

> **When to Update?**  As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutelyneeded 

Reviewer 3: ```suggestion
**When to Update?**  As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutely needed 
```

> **When to Update?**  As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutelyneeded 

Reviewer: typo: change "absolutely needed " to "absolutely needed."

Author: Updated.

>    - for **big** changes or backwards incompatible changes.

Reviewer 3: We may want to define "backwards incompatible" more precisely -- technically speaking, almost all changes to Go structs are backwards incompatible (in a not-serious way for our cases). Perhaps some language referring to json schema compatibility (adding fields is fine, removing them is not) would make this clearer. My metric is "can I apply the same resource I used to without it failing k8s validation?"

Author: Yeah, I think that sounds reasonable 👍

Author: Updated. Added a couple bullet points on backwards-incompatible changes (and when a non-breaking change may still require a version bump)

### Devfile JSON Schema

As mentioned above, the Devfile JSON schema is generated from the Go structs defined in the Devfile Kubernetes API. The latest JSON schema for a given API version is located under `schemas/<api-version>` in the [devfile/api repo](https://github.com/devfile/api/). 

> **Versioning Scheme**: Semver (major.minor.bugfix)

Reviewer: Semver: Either use the long-form or use the formal short-form `SemVer`

Author: Haha, I had debated on whether `Semver` or `SemVer` was proper. I'll actually use `Semantic Versioning` to be more clear. Updated.

**Update Process**: 
  
   1) Update the schema version string in the `// +devfile:jsonschema:version=<schema-version>` annotation in `pkg/apis/workspace/<k8s-api-version>/doc.go`
   2) Re-generate the json schema
   3) Publish new schema to devfile.io website

> **When to Update?** On each release of the schema, incremented based on the changes going in to the release (e.g. bugfix = comments / bugfixes, minor = larger and / or backwards compatible changes, major == breaking / backwards incompatible changes). K8S API version updates should also result in an appropriate increment of the schema version.

Reviewer: Consider changing to a list for easy reading. The major entry is using == instead of =.

Author: Sounds good to me, updated.

### Relationship Between K8s API version and JSON Schema Version

> The Devfile JSON schema is generated from the Kubernetes API, and the version for the JSON schema is set in the doc.go file in the K8S API (`pkg/apis/workspace/<api-version>/doc.go).

Reviewer 3: ```suggestion
The Devfile JSON schema is generated from the Kubernetes API, and the version for the JSON schema is set in the doc.go file in the K8S API (`pkg/apis/workspace/<api-version>/doc.go`).
```

> The Devfile JSON schema is generated from the Kubernetes API, and the version for the JSON schema is set in the doc.go file in the K8S API (`pkg/apis/workspace/<api-version>/doc.go).

Reviewer: Fix the path format

Author: Good catch, updated.

> As we’re only updating the K8S API version when needed, but incrementing the schema version more frequently, this means that any given API version may point to multiple, backwards-compatible, schema versions over its lifespan. The schema version under `schemas/<api-version>` in [devfile/api repo](https://github.com/devfile/api/) points to the latest JSON schema generated from the K8S API.

Reviewer: Change "...any given API version..." to "...any given K8s API version..."

Reviewer: Should "...points to the latest JSON schema generated..." be "...points to the matching JSON schema generated..." since it depends on the schema version, it may or may not be the latest.

Author: Yup, agreed. Updated.

## Release Process
The following steps outline the steps done to release a new version of the Devfile schema and publish its schemas to the devfile.io website

   1) The release engineer tasked with creating the release clones the repository (and checks out the release branch if one already exists)

   2) `make-release.sh <schema-version> <k8s-api-version>` is run:

>       i) A release branch (the name corresponding to the schema version) is created, if one does not already exist.

Reviewer 3: Does the release branch correspond to the z-stream version or the minor revision? I.e. are the release branches `v2.1.x` or `v2.1.0`, `v2.1.1`, etc.?

Author: Release branches correspond to the minor version, so `v{major}.{minor}.x`

      ii) The schema-version is updated in `pkg/apis/workspace/<api-version>/doc.go`.

      iii) New JSON schemas are generated to include the new schema version
      
      iv) A new commit including the changes

      v) Finally, a PR is opened to merge these changes into the release branch

   3) Once the release PR is approved and merged, the release engineer creates a new release on GitHub based off the release branch that was just created and includes the generated `devfile.json` as a release artifact. 

>        - The tag `v{major}.{minor}.{bugfix}`, where the semver corresponds to the devfile schema version, is used to enable the new version of the API to be pulled in as a Go module.

Reviewer: Change "semver" to "SemVer".

Reviewer: Should we tag it with `v{major}.{minor}.{bugfix}` or just `{major}.{minor}.{bugfix}` (without the starting v)?

Author: `v{major}.{minor}.{bugfix}` is used here so that the tagged release can be pulled down as a Go module (which requires `v` at the beginning of a version)

   4) Once the release is published, GitHub actions run to publish the new schema to devfile.io. The “stable” schema is also updated to point to the new schema.

An example pull request, `make-release.sh` script and GitHub action can be found here:

> - [Release Pull Request](https://github.com/johnmcollier/api/pull/7)

Reviewer: I assume these example links will need to be changed to something available in our main branch.

Author: Yup, they'll need to be updated once the release script and github workflow yaml is committed in this repo.

- [make-release.sh](https://github.com/johnmcollier/api/blob/master/make-release.sh)
- [release-schema.yaml](https://github.com/johnmcollier/api/blob/master/.github/workflows/release-schema.yaml)

> 

Reviewer: And one extra step to send an announcement to the mailing list and slack channel to broadcast the availability of the new release after the release is completed.

Author: Sounds good to me. Updated.

### Kubernetes API
The Devfile Kubernetes API (defined in https://github.com/devfile/api/) is the single source of truth for the Devfile spec. It’s defined under `pkg/apis/` and contains the Go structs that are used by the devworkspace operator and devfile library. It's also where the Devfile JSON schema is generated from. 

**Versioning Scheme**: [Kubernetes](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning) (e.g. v1alpha1, v1beta2, v1, v2, etc)

**How to Update?**: 

   1) Add a new folder for the version in the [devfile/api](https://github.com/devfile/api/) repository under [pkg/apis/workspaces](https://github.com/devfile/api/tree/master/pkg/apis/workspaces). For example `pkg/apis/workspaces/v1` if bumping the K8S API version to `v1`.
   2) Add a schema and version in the CRD manifests
   3) Go through the JSON schema update process outlined below to update the JSON schema version.
   4) Update the devworkspace operator and devfile library to consume the Go structs in the new K8S API version, as needed.

> **When to Update?**

Reviewer: Add line break after title

> **When to Update?**  As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutely needed (for **big** changes or backwards incompatible changes).

Reviewer 3: Optional: I'd throw a newline here for readability:
```suggestion
**When to Update?**
As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutely needed (for **big** changes or backwards incompatible changes).
```

>    - Backwards incompatible changes are defined as any change in the schema that would cause K8S API validation errors on the old resource version (e.g. removed fields or new mandatory fields)

Reviewer 3: ```suggestion
   - Backwards incompatible changes are defined as any change in the schema that would cause K8S API validation errors on the old resource version (e.g. removed fields or new mandatory fields without a default)
```

# Devfile Versioning and Release Process
This design document outlines the proposed versioning and release process for the Devfile spec. 

This document summarizes parts from the Devfile API technical meeting slides presented by @davidfestal back in October, and I recommend having a read through of it: https://docs.google.com/presentation/d/1ohM1HzPB59a3ajvB7rVWJkKONLBAuT0mb5P20_A6vLs/edit#slide=id.g8fc722bef9_1_8

## Versioning

The following sections outline how we version the Devfile Kubernetes API, as well as the JSON schema generated from the API.

### Kubernetes API
The Devfile Kubernetes API (defined in https://github.com/devfile/api/) is the single source of truth for the Devfile spec. It’s defined under `pkg/apis/` and contains the Go structs that are used by the devworkspace operator and devfile library. It's also where the Devfile JSON schema is generated from. 

**Versioning Scheme**: [Kubernetes](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning) (e.g. v1alpha1, v1beta2, v1, v2, etc)

> **How to Update**: 

Reviewer: Change to "How to Update?" to make it consistent with the "When to Update?"

   1) Add a new folder for the version in the [devfile/api](https://github.com/devfile/api/) repository under [pkg/apis/workspaces](https://github.com/devfile/api/tree/master/pkg/apis/workspaces). For example `pkg/apis/workspaces/v1` if bumping the K8S API version to `v1`.
   2) Add a schema and version in the CRD manifests

>    3) Generate the JSON schema from the API. New JSON schema will be located under `schemas/latest`.

Reviewer 4: Since we generate the JSON schema here, should we declare that new version should be published to devfile.io website?
If it's supposed that we MUST update `Devfile JSON Schema` each time K8s API is updated then maybe it makes sense to update this item to be `update JSON Schema` with the reference to the process described below?

Author: Yeah, agreed.

   4) Update the devworkspace operator and devfile library to consume the Go structs in the new K8S API version, as needed.

**When to Update?**
As incrementing the Kubernetes API version for Devfile is a relatively heavy process, and affects the library, only update the K8s API version when absolutely needed (for **big** changes or backwards incompatible changes).
   - Backwards incompatible changes are defined as any change in the schema that would cause K8S API validation errors on the old resource version (e.g. removed fields or new mandatory fields without a default)
   - New, optional fields to the API do not necessarily require a version bump, but if it's a large change, it may be wise to bump the version.

### Devfile JSON Schema

As mentioned above, the Devfile JSON schema is generated from the Go structs defined in the Devfile Kubernetes API. The latest JSON schema for a given K8S API version is located under `schemas/<api-version>` in the [devfile/api repo](https://github.com/devfile/api/). 

**Versioning Scheme**: Semantic Versioning (major.minor.bugfix)

> **How to Update**: 

Reviewer: Change to "How to Update?" to make it consistent with the "When to Update?"

  
   1) Update the schema version string in the `// +devfile:jsonschema:version=<schema-version>` annotation in `pkg/apis/workspace/<k8s-api-version>/doc.go`
   2) Re-generate the json schema

>    3) Publish new schema to devfile.io website

Reviewer 4: This item confuses me a bit. It makes an impression that the schema needs to be updated manually, but we are not promoting doing it, since we have(or will have) a workflow that will publish a new schema of a new release, as described above, right?
It may be just my personal feeling.

Author: Yup, that's a good point. I'll make it clear that it's the workflow that's publishing the schema

**When to Update?** On each release of the schema, incremented based on the changes going in to the release. E.g.:
   

>    - major == breaking / backwards incompatible changes). 

Reviewer: Remove the ending ")."

**Versioning Scheme**: [Kubernetes](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning) (e.g. v1alpha1, v1beta2, v1, v2, etc)

> **How to Update?**: 

Reviewer: Remove the ending :

**Versioning Scheme**: Semantic Versioning (major.minor.bugfix)

> **How to Update?**: 

Reviewer: Remove ending :

  
   1) Update the schema version string in the `// +devfile:jsonschema:version=<schema-version>` annotation in `pkg/apis/workspace/<k8s-api-version>/doc.go`
   2) Re-generate the json schema

> **When to Update?** On each release of the schema, incremented based on the changes going in to the release. E.g.:

Reviewer: Add line break after title
